service/history/ndc/history_importer.go

```diff
@@ -0,0 +1,459 @@
+			mutableStateSpec,
+			task,
+		)
+	}
+	return r.applyNonStartEventsAndSerialize(
+		ctx,
+		wfContext,
+		mutableState,
+		mutableStateSpec,
+		task,
+		createNewBranch,
+	)
+}
+
+func (r *HistoryImporterImpl) applyStartEventsAndSerialize(
+	ctx context.Context,
+	wfContext workflow.Context,
+	mutableState workflow.MutableState,
+	mutableStateSpec MutableStateInitializationSpec,
+	task replicationTask,
+) ([]byte, error) {
+	mutableState, newMutableState, err := r.mutableStateMapper.ApplyEvents(
+		ctx,
+		wfContext,
+		mutableState,
+		task,
+	)
+	if err != nil {
+		return nil, err
+	}
+	if newMutableState != nil {
+		task.getLogger().Error(
+			"HistoryImporter::applyStartEventsAndSerialize encountered create workflow with continue as new case",
+			tag.Error(err),
+		)
+	}
+	return r.persistHistoryAndSerializeMutableState(ctx, mutableState, mutableStateSpec)
+}
+
+func (r *HistoryImporterImpl) applyNonStartEventsAndSerialize(
+	ctx context.Context,
+	wfContext workflow.Context,
+	mutableState workflow.MutableState,
+	mutableStateSpec MutableStateInitializationSpec,
+	task replicationTask,
+	createNewBranch bool,
+) ([]byte, error) {
+	prepareBranchFn := r.mutableStateMapper.GetOrCreateHistoryBranch
+	if createNewBranch {
+		prepareBranchFn = r.mutableStateMapper.CreateHistoryBranch
+	}
+	mutableState, prepareHistoryBranchOut, err := prepareBranchFn(ctx, wfContext, mutableState, task)
+	if err != nil {
+		return nil, err
+	} else if !prepareHistoryBranchOut.DoContinue {
+		return r.persistHistoryAndSerializeMutableState(ctx, mutableState, mutableStateSpec)
+	} else if createNewBranch && prepareHistoryBranchOut.BranchIndex == 0 {
+		// sanity check
+		return nil, serviceerror.NewInternal("HistoryImporter unable to correctly create new branch")
+	}
+
+	mutableState, _, err = r.mutableStateMapper.GetOrRebuildMutableState(
+		ctx,
+		wfContext,
+		mutableState,
+		GetOrRebuildMutableStateIn{replicationTask: task, BranchIndex: prepareHistoryBranchOut.BranchIndex},
+	)
+	if err != nil {
+		return nil, err
+	}
+	mutableState, newMutableState, err := r.mutableStateMapper.ApplyEvents(
+		ctx,
+		wfContext,
+		mutableState,
+		task,
+	)
+	if err != nil {
+		return nil, err
+	}
+	if newMutableState != nil {
+		task.getLogger().Error(
+			"HistoryImporter::applyNonStartEventsAndSerialize encountered create workflow with continue as new case",
+			tag.Error(err),
+		)
+	}
+	return r.persistHistoryAndSerializeMutableState(ctx, mutableState, mutableStateSpec)
+}
+
+func (r *HistoryImporterImpl) persistHistoryAndSerializeMutableState(
+	ctx context.Context,
+	mutableState workflow.MutableState,
+	mutableStateSpec MutableStateInitializationSpec,
+) ([]byte, error) {
+	targetWorkflowSnapshot, targetWorkflowEventsSeq, err := mutableState.CloseTransactionAsSnapshot(
+		workflow.TransactionPolicyPassive,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	sizeSiff, err := workflow.PersistWorkflowEvents(ctx, r.shardContext, targetWorkflowEventsSeq...)
```

Comment:
will this be a problem if the history scavenger kick in?

Reply:
yes but there will be 60 days delay right?